src/t2c.c

```diff
@@ -0,0 +1,314 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <llvm-c/Analysis.h>
+#include <llvm-c/BitWriter.h>
+#include <llvm-c/Core.h>
+#include <llvm-c/ExecutionEngine.h>
+#include <llvm-c/Target.h>
+#include <llvm-c/Transforms/PassBuilder.h>
+#include <stdlib.h>
+
+#include "jit.h"
+#include "riscv_private.h"
+
+#define MAX_BLOCKS 8152
+
+struct LLVM_block_map_entry {
+    uint32_t pc;
+    LLVMBasicBlockRef block;
+};
+
+struct LLVM_block_map {
+    uint32_t count;
+    struct LLVM_block_map_entry map[MAX_BLOCKS];
+};
+
+FORCE_INLINE void LLVM_block_map_insert(struct LLVM_block_map *map,
+                                        LLVMBasicBlockRef *entry,
+                                        uint32_t pc)
+{
+    struct LLVM_block_map_entry map_entry;
+    map_entry.block = *entry;
+    map_entry.pc = pc;
+    map->map[map->count++] = map_entry;
+    return;
+}
+
+FORCE_INLINE LLVMBasicBlockRef LLVM_block_map_search(struct LLVM_block_map *map,
+                                                     uint32_t pc)
+{
+    for (uint32_t i = 0; i < map->count; i++) {
+        if (map->map[i].pc == pc) {
+            return map->map[i].block;
+        }
+    }
+    return NULL;
+}
+
+#define RVT2OP(inst, code)                                                \
+    static void t2_##inst(                                                \
+        LLVMBuilderRef *builder UNUSED, LLVMTypeRef *param_types UNUSED,  \
+        LLVMValueRef start UNUSED, LLVMBasicBlockRef *entry UNUSED,       \
+        LLVMBuilderRef *taken_builder UNUSED,                             \
+        LLVMBuilderRef *untaken_builder UNUSED, uint64_t mem_base UNUSED, \
+        rv_insn_t *ir UNUSED)                                             \
+    {                                                                     \
+        code;                                                             \
+    }
+
+FORCE_INLINE LLVMValueRef build_rs1_addr(LLVMValueRef start,
+                                         LLVMBuilderRef *builder,
+                                         rv_insn_t *ir)
+{
+    LLVMValueRef rs1_offset[1] = {LLVMConstInt(
+        LLVMInt32Type(), offsetof(riscv_t, X) / sizeof(int) + ir->rs1, true)};
+    return LLVMBuildInBoundsGEP2(*builder, LLVMInt32Type(),
+                                 LLVMGetParam(start, 0), rs1_offset, 1,
+                                 "addr_rs1");
```

Comment:
C language standard (i.e. C99 and C11) defines a predefined identifier as follows in clause 6.4.2.2:
> "The identifier `__func__` shall be implicitly declared by the translator as if, immediately following the opening brace of each function definition, the declaration
```c
static const char __func__[] = "function-name";
```
> appeared, where function-name is the name of the lexically-enclosing function."

It means that the C compiler implicitly adds `__func__` in every function so that it can be used in that function to get the function name.